File: apc-mcp/node/src/runtime/stateMachine.ts
export enum State { OK="OK", WARN="WARN", SOFT="SOFT", HARD="HARD", RESCUE="RESCUE" }

type Event = "green"|"warn"|"soft"|"hard"|"rescue"|"reset";

export function nextState(cur: State, ev: Event): State {
  switch (ev) {
    case "green":  return (cur===State.OK||cur===State.WARN) ? State.OK : cur;
    case "warn":   return cur===State.OK ? State.WARN : cur;
    case "soft":   return (cur===State.OK||cur===State.WARN) ? State.SOFT : cur;
    case "hard":   return State.HARD;
    case "rescue": return State.RESCUE;
    case "reset":  return State.OK;
    default:       return cur;
  }
}

File: agentflow/apc_runtime/state_machine.py
from enum import Enum
from typing import Literal, Tuple

class State(str, Enum):
    OK="OK"; WARN="WARN"; SOFT="SOFT"; HARD="HARD"; RESCUE="RESCUE"

Transition = Tuple[State, State, str]

def next_state(cur: State, event: Literal["green","warn","soft","hard","rescue","reset"]) -> State:
    # minimal deterministic policy
    if event == "green":
        return State.OK if cur in (State.OK, State.WARN) else cur
    if event == "warn":
        return State.WARN if cur == State.OK else cur
    if event == "soft":
        return State.SOFT if cur in (State.OK, State.WARN) else cur
    if event == "hard":
        return State.HARD
    if event == "rescue":
        return State.RESCUE
    if event == "reset":
        return State.OK
    return cur
